cdc/puller/multiplexing_puller.go

```diff
@@ -0,0 +1,435 @@
+			p.CounterKv.Inc()
+			if err := progress.consume.f(ctx, e.Val, progress.spans); err != nil {
+				return errors.Trace(err)
+			}
+		} else if e.Resolved != nil {
+			p.CounterResolved.Add(float64(len(e.Resolved.Spans)))
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case progress.resolvedEventsCache <- e:
+				p.schedule(ctx, progress)
+			default:
+				p.CounterResolvedDropped.Add(float64(len(e.Resolved.Spans)))
+			}
+		}
+	}
+}
+
+func (p *MultiplexingPuller) getProgress(subID kv.SubscriptionID) *tableProgress {
+	p.subscriptions.RLock()
+	defer p.subscriptions.RUnlock()
+	return p.subscriptions.m[subID]
+}
+
+func (p *MultiplexingPuller) getAllProgresses() map[*tableProgress]struct{} {
+	p.subscriptions.RLock()
+	defer p.subscriptions.RUnlock()
+	hashset := make(map[*tableProgress]struct{}, len(p.subscriptions.m))
+	for _, value := range p.subscriptions.m {
+		hashset[value] = struct{}{}
+	}
+	return hashset
+}
+
+func (p *MultiplexingPuller) getProgressBySpan(span tablepb.Span) *tableProgress {
+	p.subscriptions.RLock()
+	defer p.subscriptions.RUnlock()
+	return p.subscriptions.n.GetV(span)
+}
+
+func (p *MultiplexingPuller) schedule(ctx context.Context, progress *tableProgress) {
+	if progress.scheduled.CompareAndSwap(false, true) {
+		select {
+		case <-ctx.Done():
+		case p.advanceCh <- progress:
+		}
+	}
+}
+
+func (p *MultiplexingPuller) advanceSpans(ctx context.Context) error {
+	handleProgress := func(ctx context.Context, progress *tableProgress) error {
+		defer func() {
+			progress.scheduled.Store(false)
+			if len(progress.resolvedEventsCache) > 0 {
+				p.schedule(ctx, progress)
+			}
+		}()
+
+		var event kv.MultiplexingEvent
+		var spans *model.ResolvedSpans
+		for i := 0; i < 128; i++ {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case event = <-progress.resolvedEventsCache:
+				spans = event.RegionFeedEvent.Resolved
+			default:
+				return nil
+			}
+			if spans == nil {
+				// It means the event comes from `checkResolveLock`.
+				progress.resolveLock()
+				continue
+			}
+			p.queueResolvedDuration.Observe(float64(time.Since(event.Start).Milliseconds()))
+			if err := progress.handleResolvedSpans(ctx, spans); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	}
+
+	var progress *tableProgress
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case progress = <-p.advanceCh:
+			if err := handleProgress(ctx, progress); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (p *MultiplexingPuller) checkResolveLock(ctx context.Context) error {
+	resolveLockTicker := time.NewTicker(resolveLockTickInterval)
+	defer resolveLockTicker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-resolveLockTicker.C:
+		}
+		for progress := range p.getAllProgresses() {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case progress.resolvedEventsCache <- kv.MultiplexingEvent{}:
```

Comment:
Why not call `progress.resolveLock` directly?

Reply:
Good idea.